README.markdown
What is JSIO? [![Build Status](https://travis-ci.org/alanshaw/JavaScript-Image-Optimiser.png)](https://travis-ci.org/alanshaw/JavaScript-Image-Optimiser)
=============

JSIO is a tiny library that allows you to make fewer requests to your server by packaging all your site image data in a JavaScript file in [data uri format](http://en.wikipedia.org/wiki/Data_URI_scheme). The idea is that a Gzipped JavaScript file containing all your image data is likely to be smaller than the sum of each image file and its associated HTTP request and header traffic.

Huh?
----

Each background-image you place in your CSS stylesheets and each `<img>` element you place in your HTML file causes your browser to issue a request to download the file from the web server. Each request requires a new HTTP connection to be setup, and HTTP headers to be sent down the wire before the file even starts to be transferred. With a lot of images this starts to become a problem, as a lot of unnecessary time and bandwidth is wasted performing these basic operations over and over again. JSIO helps to ease the burden of opening all these connections by placing all your image data in one file.


How does it work?
=================

Simple. The JSIO script sends a request to the server to fetch a JavaScript file that contains image data. The data is loaded into the jsio.resources object - the keys are the filenames of images and the values are their data uri representation. For example:

```javascript
jsio.resources = {'superman.jpg': 'data:image/jpeg;base64,/9j/4AAQSkZJRgABAQAAAQABAAD/2wCEAAkGBhISEBU...'};
```

Check this out:

```html
<img src="img/jsio.gif#superman.jpg"/>
```

This is an image element that will display the image &quot;jsio.gif&quot;. The file &quot;jsio.gif&quot; is a placeholder that signals to the library that the image by the name of &quot;img/superman.jpg&quot; has been encoded as a data URI and can be found in the jsio.resources JavaScript object.

JSIO looks for these placeholders, and replaces them with data uri's that are in the jsio.resources object.

```html
<img src="data:image/jpeg;base64,/9j/4AAQSkZJRgABAQAAAQABAAD/2wCEAAkGBhISEBU..."/>
```

You can see JSIO in action on the [test page](http://alanshaw.github.com/JavaScript-Image-Optimiser/test.html).

What's the catch?
=================

Glad you asked. The elephant in the room is of course IE. IE < 9 has a 32KB limit on data uri's and IE < 8 doesn't support data uri's at all. [Data uri's are supported in most other browsers](http://en.wikipedia.org/wiki/Data_URI_scheme#Web_browser_support). Also, somewhat obviously, JavaScript must be enabled.

How can I use it on my website?
===============================

- Clone the latest [JSIO code](https://github.com/alanshaw/JavaScript-Image-Optimiser/)
- Link to the JSIO library _at the bottom_ of your document: `<script src="js/jsio.js" data-res-url="/js/jsio-resources.js"></script>`
- Place the placeholder image (jsio.gif) in your website image directory
- [Use this tool to generate your resources file](http://alanshaw.github.com/JavaScript-Image-Optimiser/to-data-url.html). There is also a [grunt task](https://npmjs.org/package/grunt-jsio) that'll do this for you.
- Place the resources file at "/js/jsio-resources.js"
- Start using JSIO URLs in your `<img>` elements and background-image CSS declarations

[Advanced usage](https://github.com/alanshaw/JavaScript-Image-Optimiser/blob/master/doc/advanced-usage.md)

What's new in this version?
===========================

See the [changelog](https://github.com/alanshaw/JavaScript-Image-Optimiser/blob/master/CHANGELOG)

I have a question/suggestion/problem...
=======================================

Please [contact me](http://freestyle-developments.co.uk/contact) or [add an issue on GitHub](https://github.com/alanshaw/JavaScript-Image-Optimiser/issues)

...or check the [frequently asked questions](https://github.com/alanshaw/JavaScript-Image-Optimiser/blob/master/doc/faq.md)

CHANGELOG
v1.1.1
======

- Fixed problem processing images where path to image file wouldn't be prefixed to the filename when the file is not found in the JSIO resources file
- Added quotes around background image URLs when we can't use a data URL incase the URL contains special characters.
- Added parameter to process function that allows the callee to control whether the 404 fallback occurs or not
- Fixed error thrown whilst trying to process cross domain stylesheets (can't be done so now skips over them)

v1.1.0
======

- Exposed "process" function to allow new image elements and style rules with JSIO URLs to be processed
- "process" function optionally takes a jsio resources object parameter
- Use HTML5 "data-res-url" attribute instead of querystring to specify jsio resources file URL
- If no resources URL is specified, don't process jsio URLs - assume the user is going to manually call "process"
- General preformance tweaks
- Moved to grunt for build and test

v1.0.0
======

- Added IE < 8 optimisation (IE < 8 does not download the jsio-resources.js)
- Opimisations for minified (but not gzipped) script
- Fixed problem with getQueryString function that wouldn't return a default value if query string key present but not value  

v1.0.0 Beta
===========

- Refactored code into two "processing" functions - one for CSS background-images and one for image elements
- Added IE detection script (thanks to https://gist.github.com/527683)
- Fixed CSS background image processing function (now works in IE)
- Added rules for IE < 8 to remove placeholder and expose fallback image for all images and background-images
- Added rules for IE 8 to remove placeholder and expose fallback image when data uri exceeds 32KB
- Improved regular expressions that look for jsio placeholder gif for both processing functions
- Added fix that makes IE < 9 fire callback function when jsio resources file finishes loading

v1.0.0 Alpha
============

- Tout le monde

test/jsio.js
(function() {
	
	"use strict";
	
	var fs = require('fs');
	var vm = require('vm');
	
	var js = fs.readFileSync('src/jsio.js');
	
	/**
	 * Create a mock document object that'll allow JSIO initialisation to complete successfully (without auto loading a
	 * jsio resources file).
	 */
	function mockDocument() {
		return {
			
			getElementsByTagName: function(tagName) {
				
				// For when JSIO tries to get it's <script> tag
				if(tagName == 'script') {
					return [{
						getAttribute: function() {
							return null;
						}
					}];
				}
				
				return [];
			},
			
			createElement: function() {
				return {
					getElementsByTagName: function() {
						return [];
					}
				};
			}
		};
	}
	
	function runInContext(initContext) {
		
		var context = vm.createContext(initContext);
		
		vm.runInContext(js, context);
		
		return context;
	}
	
	module.exports = {
		
		testJsioNamespaceDefined: function(test) {
			
			var context = runInContext({document: mockDocument()});
			
			test.expect(1);
			
			test.notEqual(context.jsio, null, 'Should contain JSIO namespace');
			
			test.done();
		},
		
		testProcessFunctionExported: function(test) {
			
			var context = runInContext({document: mockDocument()});
			
			test.expect(1);
			
			test.notEqual(context.jsio.process, null, 'Process function should be exported');
			
			test.done();
		},
		
		testAutoProcess: function(test) {
			
			var resScript = null;
			var resUrl = 'path/to/resources.js';
			
			var document = {
			
				getElementsByTagName: function(tagName) {
					
					// For when JSIO tries to get it's <script> tag
					if(tagName == 'script') {
						return [{
							getAttribute: function(name) {
								if(name == 'data-res-url') {
									// Return the res url when getting the "data-res-url"
									return resUrl;
								}
								return null;
							},
							parentNode: {
								appendChild: function(child){}
							}
						}];
					}
					
					return [];
				},
				
				createElement: function(tagName) {
					
					var el = {
						getElementsByTagName: function() {
							return [];
						}
					};
					
					// For when JSIO creates a script tag to get the jsio resources file
					if(tagName == 'script') {
						// Store the element so we can interrogate it later
						resScript = el;
					}
					
					return el;
				}
			};
			
			runInContext({document: document});
			
			test.expect(3);
			
			test.notEqual(resScript, null, 'Resources script element should have been stored for interrogation');
			
			test.strictEqual(resScript.src, resUrl, 'Resources script element should have expected src');
			
			test.notEqual(resScript.onload, undefined, 'An onload function should have been registered with the resources script');
			
			test.done();
		},
		
		testIe7AutoProcess: function(test) {
			
			var resScript = null;
			var resUrl = 'path/to/resources.js';
			var getElementsByTagNameCount = 0;
			
			var document = {
			
				getElementsByTagName: function(tagName) {
					
					// For when JSIO tries to get it's <script> tag
					if(tagName == 'script') {
						return [{
							getAttribute: function(name) {
								if(name == 'data-res-url') {
									// Return the res url when getting the "data-res-url"
									return resUrl;
								}
								return null;
							}
						}];
					}
					
					return [];
				},
				
				createElement: function(tagName) {
					
					var el = {
						getElementsByTagName: function() {
							return [];
						}
					};
					
					// For when JSIO creates a script tag to get the jsio resources file
					if(tagName == 'script') {
						
						// Store the element so we can interrogate it later
						resScript = el;
						
					} else if(tagName == 'div') {
						
						return {
							getElementsByTagName: function() {
								
								getElementsByTagNameCount++;
								
								// Return a list of 1 element so we can bump the ie variable up to 7
								if(getElementsByTagNameCount < 4) return [true];
								
								return [];
							}
						};
					}
					
					return el;
				},
				
				styleSheets: []
			};
			
			var context = runInContext({document: document});
			
			test.expect(2);
			
			test.strictEqual(resScript, null, 'No resources script element should have been created');
			
			test.notEqual(context.jsio.resources, undefined, 'A dummy jsio.resources object should have been created');
			
			test.done();
		},
		
		testIe8AutoProcess: function(test) {
			
			var resScript = null;
			var resUrl = 'path/to/resources.js';
			var getElementsByTagNameCount = 0;
			
			var document = {
			
				getElementsByTagName: function(tagName) {
					
					// For when JSIO tries to get it's <script> tag
					if(tagName == 'script') {
						return [{
							getAttribute: function(name) {
								if(name == 'data-res-url') {
									// Return the res url when getting the "data-res-url"
									return resUrl;
								}
								return null;
							},
							parentNode: {
								appendChild: function(child){}
							}
						}];
					}
					
					return [];
				},
				
				createElement: function(tagName) {
					
					var el = {
						getElementsByTagName: function() {
							return [];
						}
					};
					
					// For when JSIO creates a script tag to get the jsio resources file
					if(tagName == 'script') {
						
						// Store the element so we can interrogate it later
						resScript = el;
						
					} else if(tagName == 'div') {
						
						return {
							getElementsByTagName: function() {
								
								getElementsByTagNameCount++;
								
								// Return a list of 1 element so we can bump the ie variable up to 8
								if(getElementsByTagNameCount < 5) return [true];
								
								return [];
							}
						};
					}
					
					return el;
				}
			};
			
			runInContext({document: document});
			
			test.expect(3);
			
			test.notEqual(resScript, null, 'Resources script element should have been stored for interrogation');
			
			test.strictEqual(resScript.src, resUrl, 'Resources script element should have expected src');
			
			test.notEqual(resScript.onreadystatechange, undefined, 'An onreadystatechange function should have been registered with the resources script');
			
			test.done();
		},
		
		testProcessCssBackgroundImages: function(test) {
			
			var document = mockDocument();
			
			// Create mock CSS data to process
			document.styleSheets = [{
				cssRules: [{
					style: {
						backgroundImage: 'url(path/to/images/jsio.gif#foo.jpg)'
					}
				}, {
					style: {
						backgroundImage: 'url("jsio.gif#bar.gif")'
					}
				}]
			}, {
				cssRules: [{
					style: {
						backgroundImage: "url('some/path/to/jsio.gif#not-found.png')"
					}
				}]
			}];
			
			var resources = {
				'foo.jpg': 'data:image/jpeg;base64,/9j/4AAQSkZJRgABAQAAAQABAAD/2wCEAAkGBhISEBU...',
				'bar.gif': 'data:image/gif;base64,R0lGODlhqQGpAfcAABMVFhQVFhUVFhUWFhQWFhYWFhcX...'
			};
			
			var context = runInContext({document: document});
			
			context.jsio.process(resources);
			
			test.expect(3);
			
			test.equal(
				document.styleSheets[0].cssRules[0].style.backgroundImage,
				'url(' + resources['foo.jpg'] + ')',
				'Should have replace backgroundImage URL for foo.jpg with data URL from JSIO resources'
			);
			
			test.equal(
				document.styleSheets[0].cssRules[1].style.backgroundImage,
				'url(' + resources['bar.gif'] + ')',
				'Should have replace backgroundImage URL for bar.gif with data URL from JSIO resources'
			);
			
			test.equal(
				document.styleSheets[1].cssRules[0].style.backgroundImage,
				"url('some/path/to/not-found.png')",
				'Should have replace backgroundImage URL for not-found.png with image URL since it was not present in JSIO resources file'
			);
			
			test.done();
		},
		
		testProcessImageElements: function(test) {
			
			var document = mockDocument();
			
			document.styleSheets = [];
			
			var imageElements =[{
				src: 'path/to/images/jsio.gif#foo.jpg'
			}, {
				src: 'path/to/images/jsio.gif#bar.gif'
			}, {
				src: 'path/to/images/jsio.gif#not-found.png'
			}];
			
			document.getElementsByTagName = function(tagName) {
				
				// For when JSIO tries to get it's <script> tag
				if(tagName == 'script') {
					return [{
						getAttribute: function() {
							return null;
						}
					}];
					
				} else if(tagName == 'img') {
					return imageElements;
				}
				
				return [];
			};
			
			var resources = {
				'foo.jpg': 'data:image/jpeg;base64,/9j/4AAQSkZJRgABAQAAAQABAAD/2wCEAAkGBhISEBU...',
				'bar.gif': 'data:image/gif;base64,R0lGODlhqQGpAfcAABMVFhQVFhUVFhUWFhQWFhYWFhcX...'
			};
			
			var context = runInContext({document: document});
			
			context.jsio.process(resources);
			
			test.expect(3);
			
			test.equal(
				imageElements[0].src,
				resources['foo.jpg'],
				'Should have replace image src for foo.jpg with data URL from JSIO resources'
			);
			
			test.equal(
				imageElements[1].src,
				resources['bar.gif'],
				'Should have replace image src for bar.gif with data URL from JSIO resources'
			);
			
			test.equal(
				imageElements[2].src,
				'path/to/images/not-found.png',
				'Should have replace image src for not-found.png with image URL since it was not present in JSIO resources file'
			);
			
			test.done();
		}
	};
}());

js/script.js
$(function(){
	
	var resources = $('#resources');
	
	if (window.File && window.FileReader && window.FileList && window.Blob) {
		
		var jsioResources = $('#jsio-resources');
		
		resources.change(function(evt) {
			
			var files = evt.target.files,
				flen = files.length,
				unreadFiles = flen,
				dataUrls = [];
			
			for(var i = 0; i < flen; ++i) {
				readFile(files[i]);
			}
			
			function readFile(file) {
				
				var reader = new FileReader();
				
				reader.onload = function(evt) {
					
					dataUrls.push('"' + file.name + '":"' + reader.result + '"');
					unreadFiles = unreadFiles - 1;
					
					if(unreadFiles == 0) {
						jsioResources.val('jsio.resources={' + dataUrls.join(',') + '}');
					}
				};
				
				reader.readAsDataURL(file);
			}
		});
		
		$('#select-all').click(function() {
			jsioResources[0].select();
		});
		
	} else {
		
		if(resources.length)
			alert('Can\'t generate JSIO resources file: The File APIs are not fully supported in this browser.');
	}
});


js/jsio-1.0.0-beta.js
/**
 * JSIO v1.0.0 Beta
 * 
 * Copyright 2011 Alan Shaw
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and limitations
 * under the License.
 */
(function(w, d) {
	
	// Init ///////////////////////////////////////////////////////////////////
	
	// Setup the JSIO namespace
	if(!w.jsio) w.jsio = {};
	
	var scripts = d.getElementsByTagName('script'),
		jsioScript = scripts[scripts.length - 1],
		qs = '?' + jsioScript.src.replace(/^[^\?]+\??/, ''),
		// Check for IE thanks to https://gist.github.com/527683
		ie = (function(){
		
			var undef, v = 3, div = d.createElement('div');
			
			// the while loop is used without an associated block: {}
			// so, only the condition within the () is executed.
		
			// semicolons arent allowed within the condition,
			//   so a comma is used to stand in for one
			// basically allowing the two separate statements 
			//   to be evaluated sequentially.
			
			while (
				div.innerHTML = '<!--[if gt IE '+(++v)+']><i></i><![endif]-->',
				div.getElementsByTagName('i')[0]
			);
			
			// each time it's evaluated, v gets incremented and
			//   tossed into the DOM as a conditional comment
			// the i element is then a child of the div.
			
			// the return value of the getEBTN call is used as 
			//   the final condition expression
			// if there is an i element (the IE conditional
			//   succeeded), then getEBTN's return is truthy
			// and the loop continues until there is no 
			//   more i elements.
			
			// In other words:  ** MAGIC**
			
			return v > 4 ? v : undef;
			
		}());
	
	// Function ///////////////////////////////////////////////////////////////
	
	function getQueryString(key, dflt) {
		
		if(!dflt) dflt = '';
		
		var regex = new RegExp("[\\?&]" + key + "=([^&#]*)"),
			match = regex.exec(qs);
		
		return match ? match[1] : dflt;
	}
	
	function processCssBackgroundImages() {
		
		// CSS background images
		for(var i = 0, ilen = d.styleSheets.length; i < ilen; ++i) {
			
			var styleSheet = d.styleSheets[i],
				// Allows access to the non-standard list of CSS rules in IE
				cssRules = ie < 9 ? 'rules' : 'cssRules';
			
			for(var j = 0, jlen = styleSheet[cssRules].length; j < jlen; ++j) {
				
				var rule = styleSheet[cssRules][j],
					bgImageStyle = rule.style.backgroundImage;
				
				if(bgImageStyle) {
					
					var regex = /url\(['"]?(.*?)jsio\.gif#(.*?)['"]?\)/gmi,
						match,
						dataUriBgImageStyle = bgImageStyle;
					
					while(match = regex.exec(bgImageStyle)) {
						
						if(jsio.resources[match[2]]) {
							
							// IE < 8 does not support data uri and IE8 has a 32KB limit
							if(ie < 8 || (ie == 8 && jsio.resources[match[2]].length * 2 > 32000)) {
								
								dataUriBgImageStyle = dataUriBgImageStyle.replace(match[0], 'url(' + match[1] + match[2] + ')');
								
							} else {
								
								dataUriBgImageStyle = dataUriBgImageStyle.replace(match[0], 'url(' + jsio.resources[match[2]] + ')');
							}
							
						} else {
							
							dataUriBgImageStyle = dataUriBgImageStyle.replace(match[0], 'url(' + match[1] + match[2] + ')');
						}
					}
					
					if(dataUriBgImageStyle != bgImageStyle) {
						rule.style.backgroundImage = dataUriBgImageStyle;
					}
				}
			}
			
			// IE doesn't re-parse stylesheet data when manipulating existing rules
			if(ie) styleSheet.addRule('jsio', 'jsio');
		}
	}
	
	function processImageElements() {
		
		var images = d.getElementsByTagName('img');
		
		for(var i = 0, ilen = images.length; i < ilen; ++i) {
			
			var regex = /(.*?)jsio\.gif#(.*)/gmi, match;
			
			if(match = regex.exec(images[i].src)) {
				
				if(jsio.resources[match[2]]) {
					
					// IE < 8 does not support data uri and IE8 has a 32KB limit
					if(ie < 8 || (ie == 8 && jsio.resources[match[2]].length * 2 > 32000)) {
						
						images[i].src = match[1] + match[2];
						
					} else {
						
						images[i].src = jsio.resources[match[2]];
					}
					
				} else {
					
					images[i].src = match[2];
				}
			}
		}
	}
	
	// Main ///////////////////////////////////////////////////////////////////
	
	function process() {
		
		// Fail fast for invalid resource files or 404 (in IE's case)
		if(!jsio.resources) return;
		
		processCssBackgroundImages();
		
		processImageElements();
	}
	
	var resScript = d.createElement('script');
	
	// IE < 9 does not fire onload callback on script elements
	if(ie < 9) {
		
		resScript.onreadystatechange = function() {
			
			if(this.readyState == 'loaded' || this.readyState == 'complete') {
				
				process();
			}
		}
		
	} else {
		
		resScript.onload = process;
	}
	
	jsioScript.parentNode.appendChild(resScript, jsioScript);
	resScript.src = getQueryString('resUrl', '/js/jsio-resources.js');
	
})(window, document);

js/jsio-1.0.0-alpha.js
/**
 * JSIO v1.0.0 Alpha
 * 
 * Copyright 2011 Alan Shaw
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and limitations
 * under the License.
 */
(function(w, d) {
	
	if(!w.jsio) w.jsio = {};
	
	var scripts = d.getElementsByTagName('script'),
		jsioScript = scripts[scripts.length - 1],
		qs = '?' + jsioScript.src.replace(/^[^\?]+\??/, '');
	
	function getQueryString(key, dflt) {
		
		if(!dflt) dflt = '';
		
		var regex = new RegExp("[\\?&]" + key + "=([^&#]*)"),
			match = regex.exec(qs);
		
		return match ? match[1] : dflt;
	}
	
	var resScript = d.createElement('script');
	
	resScript.onload = function() {
		
		// CSS background images
		for(var i = 0, ilen = d.styleSheets.length; i < ilen; ++i) {
			
			var styleSheet = d.styleSheets[i];
			
			for(var j = 0, jlen = styleSheet.cssRules.length; j < jlen; ++j) {
				
				var rule = styleSheet.cssRules[j],
					bgImageStyle = rule.style.getPropertyValue('background-image');
				
				if(bgImageStyle) {
					
					var regex = /url\(.*?jsio\.gif#(.*?)['"]?\)/gmi,
						match,
						dataUriBgImageStyle = bgImageStyle;
					
					while(match = regex.exec(bgImageStyle)) {
						
						if(jsio.resources[match[1]]) { // TODO: AND !(IE8 && resource > 32kb)
							dataUriBgImageStyle = dataUriBgImageStyle.replace(match[0], 'url(' + jsio.resources[match[1]] + ')');
						} else {
							dataUriBgImageStyle = dataUriBgImageStyle.replace(match[0], 'url(' + match[1] + ')');
						}
					}
					
					if(dataUriBgImageStyle != bgImageStyle) {
						rule.style.setProperty('background-image', dataUriBgImageStyle, '');
					}
				}
			}
		}
		
		// <img> elements
		var images = d.getElementsByTagName('img');
		
		for(var i = 0, ilen = images.length; i < ilen; ++i) {
			
			var regex = /jsio\.gif#(.*)/gmi, match;
			
			if(match = regex.exec(images[i].src)) {
				
				if(jsio.resources[match[1]]) { // TODO: AND !(IE8 && resource > 32kb)
					images[i].src = jsio.resources[match[1]];
				} else {
					images[i].src = match[1];
				}
			}
		}
	};
	
	jsioScript.parentNode.appendChild(resScript, jsioScript);
	resScript.src = getQueryString('resUrl', '/js/jsio-resources.js');
	
})(window, document);

src/jsio.js
(function(w, d) {
	
	// Init ///////////////////////////////////////////////////////////////////
	
	// Setup the JSIO namespace
	w.jsio = w.jsio || {};
	
	var scripts = d.getElementsByTagName('script'),
		jsioScript = scripts[scripts.length - 1],
		// Check for IE thanks to https://gist.github.com/527683
		ie = (function(){
		
			var undef, v = 3, div = d.createElement('div');
			
			// the while loop is used without an associated block: {}
			// so, only the condition within the () is executed.
		
			// semicolons arent allowed within the condition,
			//   so a comma is used to stand in for one
			// basically allowing the two separate statements 
			//   to be evaluated sequentially.
			
			while (
				div.innerHTML = '<!--[if gt IE '+(++v)+']><i></i><![endif]-->',
				div.getElementsByTagName('i')[0]
			);
			
			// each time it's evaluated, v gets incremented and
			//   tossed into the DOM as a conditional comment
			// the i element is then a child of the div.
			
			// the return value of the getEBTN call is used as 
			//   the final condition expression
			// if there is an i element (the IE conditional
			//   succeeded), then getEBTN's return is truthy
			// and the loop continues until there is no 
			//   more i elements.
			
			// In other words:  ** MAGIC**
			
			return v > 4 ? v : undef;
			
		}());
	
	// Function ///////////////////////////////////////////////////////////////
	
	function processCssBackgroundImages(resources, resource404FallbackEnabled) {
		
		// CSS background images
		for(var i = 0, ilen = d.styleSheets.length; i < ilen; ++i) {
			
			var styleSheet = d.styleSheets[i],
				// Allows access to the non-standard list of CSS rules in IE
				cssRules = ie < 9 ? 'rules' : 'cssRules',
				rulesLength;
			
			// Skip over styles in cross domain stylesheets
			// Cannot parse as SecurityError will be thrown if try to access cssRules for a cross domain stylesheet
			try {
				rulesLength = styleSheet[cssRules].length;
			} catch(e) {
				continue;
			}
			
			for(var j = 0; j <  rulesLength; ++j) {
				
				var rule = styleSheet[cssRules][j],
					bgImageStyle = rule.style.backgroundImage;
				
				if(bgImageStyle) {
					
					var regex = /url\(['"]?(.*?)jsio\.gif#(.*?)['"]?\)/gmi,
						matches = regex.exec(bgImageStyle),
						dataUriBgImageStyle = bgImageStyle;
					
					while(matches) {
						
						var fullMatch = matches[0],
							path = matches[1], 
							filename = matches[2];
						
						if(resources[filename]) {
							
							// IE < 8 does not support data uri and IE8 has a 32KB limit
							if(ie < 8 || (ie == 8 && resources[filename].length * 2 > 32000)) {
								
								dataUriBgImageStyle = dataUriBgImageStyle.replace(fullMatch, 'url("' + path + filename + '")');
								
							} else {
								
								dataUriBgImageStyle = dataUriBgImageStyle.replace(fullMatch, 'url(' + resources[filename] + ')');
							}
							
						} else if(resource404FallbackEnabled) {
							
							dataUriBgImageStyle = dataUriBgImageStyle.replace('jsio.gif#', '');
						}
						
						matches = regex.exec(bgImageStyle);
					}
					
					if(dataUriBgImageStyle != bgImageStyle) {
						rule.style.backgroundImage = dataUriBgImageStyle;
					}
				}
			}
			
			// IE doesn't re-parse stylesheet data when manipulating existing rules
			if(ie) styleSheet.addRule('jsio', 'jsio');
		}
	}
	
	function processImageElements(resources, resource404FallbackEnabled) {
		
		var images = d.getElementsByTagName('img');
		
		for(var i = 0, ilen = images.length; i < ilen; ++i) {
			
			var image = images[i],
				regex = /(.*?)jsio\.gif#(.*)/gmi, 
				matches = regex.exec(image.src);
			
			if(matches) {
				
				var path = matches[1],
					filename = matches[2], 
					src = null;
				
				if(resources[filename]) {
					
					// IE < 8 does not support data uri and IE8 has a 32KB limit
					if(ie < 8 || (ie == 8 && resources[filename].length * 2 > 32000)) {
						
						src = path + filename;
						
					} else {
						
						src = resources[filename];
					}
					
				} else if(resource404FallbackEnabled) {
					
					src = image.src.replace('jsio.gif#', '');
				}
				
				if(src) image.src = src;
			}
		}
	}
	
	// Main ///////////////////////////////////////////////////////////////////

	/**
	 * Process JSIO URLs
	 * 
	 * @param {Object} [resources] The resources for substitution. Default: window.jsio.resources.
	 * @param {Boolean} [resource404FallbackEnabled] If the resource is not found in the resources object, remove 
	 * "jsio.gif#" from the image src or backgroud image url(). Default true.
	 */
	function process(resources, resource404FallbackEnabled) {
		
		resources = resources || w.jsio.resources;
		
		// Fail fast for invalid resource files or 404 (in IE's case)
		if(!resources) return;
		
		if(resource404FallbackEnabled !== false) {
			resource404FallbackEnabled = true;
		}
		
		processCssBackgroundImages(resources, resource404FallbackEnabled);
		
		processImageElements(resources, resource404FallbackEnabled);
	}
	
	// Expose the process function so that it can be called when new image elements and style rules are added
	w.jsio.process = process;
	
	var resUrl = jsioScript.getAttribute('data-res-url');
	
	// Don't process jsio URLs if no resources file specified, assume user will call "process"
	if(!resUrl) return;
	
	// IE 7 and below can't display data uri encoded images - don't even bother to download the resources file
	if(ie < 8) {
		
		w.jsio.resources = {};
		
		process();
		
	} else {
		
		var resScript = d.createElement('script');
		
		// IE < 9 does not fire onload callback on script elements
		if(ie < 9) {
			
			resScript.onreadystatechange = function() {
				
				var state = this.readyState;
				
				if(state == 'loaded' || state == 'complete') {
					process();
				}
			};
			
		} else {
			
			resScript.onload = function() {
				process();
			};
		}
		
		jsioScript.parentNode.appendChild(resScript);
		resScript.src = resUrl;
	}
	
})(this, document);